### ISSUE TITLE
tf.gather causes reshape error during gradient computation

### ISSUE THREAD

**Issue by paradite** [tensorflow/tfjs#7494]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N.A.
- TensorFlow.js installed from (npm or script link): npm
- TensorFlow.js version (use command below): 4.2.0
- Browser version: N.A.
- Tensorflow.js Converter Version: N.A.

**Describe the current behavior**

When using `tf.gather`, the gradient computation throws an error like this:
```
Error: Invalid TF_Status: 3
Message: Input to reshape is a tensor with 64 values, but the requested shape has 4096
```

64 here is the batch size and 46*46=4096.

The reshape error occurred in [gatherGradConfig](https://github.com/tensorflow/tfjs/blob/087a02a2848db8ba1950c6894fc76d8be3936f11/tfjs-core/src/gradients/GatherV2_grad.ts#L27).

Replacing `tf.gather` usage with my own custom implementation worked and the gradient computes successfully using the tensor.

**Code where the error occurs:**

> input is tensor of shape: [ 64, 2 ], indices is tensor of shape [ 64, 1 ]

```ts
// input: Tensor {
//   shape: [ 64, 2 ],
//   dtype: 'float32',
//   size: 128,
//   strides: [ 2 ],
//   rankType: '2',
// }
// indices: Tensor {
//   shape: [ 64, 1 ],
//   dtype: 'int32',
//   size: 64,
//   strides: [ 1 ],
//   rankType: '2',
// }

const result = tf.gather(input, indices, 1, 1);
// I also tried the following, same error
// const result = tf.gather(input, indices.arraySync(), 1, 1);

// result shape: [ 64, 1 ]

const grads = optimizer.computeGradients(() => {
  // computation using result
}

// Error: Invalid TF_Status: 3
// Message: Input to reshape is a tensor with 64 values, but the requested shape has 4096
```

**Replacing with custom tf.gather implementation does not result in an error, and the gradient computation is behaving as expected:**

```ts
function gatherOwn(input: tf.Tensor, indices: tf.Tensor): tf.Tensor {
  const batchSize = input.shape[0];
  const tensors = input.split(batchSize, 0);
  const output = tf.stack(tensors.map((tensor, i) => tensor.squeeze().gather(indices.arraySync()[i])));
  return output;
}

const result = gatherOwn(input, indices);

const grads = optimizer.computeGradients(() => {
  // computation using result
}

// No error, gradient computes as expected
```

**Describe the expected behavior**

Using tf.gather should not cause the gradient computation to throw an error.

**Standalone code to reproduce the issue**
I can provide a standalone code, but it will take some time on my side.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

N.A.

**Comment by gaikwadrahul8** [comment]
Hi, @paradite 

Apologize for the delayed response and as you mentioned in issue template you'll provide standalone code or code snippet to reproduce the issue so if possible could you please help us with it ? I see you have workaround to handle tf.gather causes reshape error during gradient computation then PR will be welcomed from your end to resolve this issue. Thank you!

**Comment by POPOBE97** [comment]
> **System information**
> 
> * Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js):
> * OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS
> * Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N.A.
> * TensorFlow.js installed from (npm or script link): npm
> * TensorFlow.js version (use command below): 4.2.0
> * Browser version: N.A.
> * Tensorflow.js Converter Version: N.A.
> 
> **Describe the current behavior**
> 
> When using `tf.gather`, the gradient computation throws an error like this:
> 
> ```
> Error: Invalid TF_Status: 3
> Message: Input to reshape is a tensor with 64 values, but the requested shape has 4096
> ```
> 
> 64 here is the batch size and 46*46=4096.
> 
> The reshape error occurred in [gatherGradConfig](https://github.com/tensorflow/tfjs/blob/087a02a2848db8ba1950c6894fc76d8be3936f11/tfjs-core/src/gradients/GatherV2_grad.ts#L27).
> 
> Replacing `tf.gather` usage with my own custom implementation worked and the gradient computes successfully using the tensor.
> 
> **Code where the error occurs:**
> 
> > input is tensor of shape: [ 64, 2 ], indices is tensor of shape [ 64, 1 ]
> 
> ```ts
> // input: Tensor {
> //   shape: [ 64, 2 ],
> //   dtype: 'float32',
> //   size: 128,
> //   strides: [ 2 ],
> //   rankType: '2',
> // }
> // indices: Tensor {
> //   shape: [ 64, 1 ],
> //   dtype: 'int32',
> //   size: 64,
> //   strides: [ 1 ],
> //   rankType: '2',
> // }
> 
> const result = tf.gather(input, indices, 1, 1);
> // I also tried the following, same error
> // const result = tf.gather(input, indices.arraySync(), 1, 1);
> 
> // result shape: [ 64, 1 ]
> 
> const grads = optimizer.computeGradients(() => {
>   // computation using result
> }
> 
> // Error: Invalid TF_Status: 3
> // Message: Input to reshape is a tensor with 64 values, but the requested shape has 4096
> ```
> 
> **Replacing with custom tf.gather implementation does not result in an error, and the gradient computation is behaving as expected:**
> 
> ```ts
> function gatherOwn(input: tf.Tensor, indices: tf.Tensor): tf.Tensor {
>   const batchSize = input.shape[0];
>   const tensors = input.split(batchSize, 0);
>   const output = tf.stack(tensors.map((tensor, i) => tensor.squeeze().gather(indices.arraySync()[i])));
>   return output;
> }
> 
> const result = gatherOwn(input, indices);
> 
> const grads = optimizer.computeGradients(() => {
>   // computation using result
> }
> 
> // No error, gradient computes as expected
> ```
> 
> **Describe the expected behavior**
> 
> Using tf.gather should not cause the gradient computation to throw an error.
> 
> **Standalone code to reproduce the issue** I can provide a standalone code, but it will take some time on my side.
> 
> **Other info / logs** Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
> 
> N.A.

Thank you. You saved my day.

**Comment by POPOBE97** [comment]
Same Here. Below is the code to reproduce this issue.

```
const w = tf.variable(tf.tensor([[0, 1],
                                 [1, 2],
                                 [2, 3],
                                 [3, 4]]), true)
const y = tf.tensor([0, 1, 0, 1], [4], 'int32')

const loss = () => {
  return tf.tidy(() => {
    return tf.gather(w, y, 1, 1).sum().asScalar()
  })
}

const grad = tf.variableGrads(loss)
```

Error from console:
```
caught (in promise) Error: Size(4) must match the product of shape 4,4
    at Object.inferFromImplicitShape (util_base.ts:357:13)
    at Object.reshape [as kernelFunc] (Reshape.ts:30:23)
    at kernelFunc (engine.ts:647:22)
    at engine.ts:713:23
    at Engine.scopedRun (engine.ts:480:19)
    at Engine.runKernelFunc (engine.ts:709:10)
    at Engine.runKernel (engine.ts:554:17)
    at reshape_ (reshape.ts:60:17)
    at reshape__op (operation.ts:51:22)
    at Object.derX [as x] (GatherV2_grad.ts:51:22)
```

**Comment by nybblr** [comment]
Running into this issue as well in a network that uses `.gather()`. Seems to happen for both tfjs-core backend and tfjs-node backend. `.gather()` works fine, until I try to evaluate the gradient, same as @POPOBE97's code.

It's still slow, but here's a slightly faster version of @paradite's `gatherOwn()` workaround (also assumes axis=1, batchDims=1). Might get a bit more speed using a TensorBuffer too?

```javascript
function gatherOwn(input, indices) {
  const tensors = input.unstack(0);
  const indicesSync = indices.arraySync();
  const output = tf.stack(tensors.map((tensor, i) => tensor.gather(indicesSync[i])));
  return output;
}
```

**Comment by paradite** [comment]
@gaikwadrahul8 sorry for the late reply. I haven't had time to do proper reproduction, but I see @POPOBE97 has provided a minimal reproduction already in the comment.

Given that multiple people have reported this issue, should it be reopened? I can't promise to submit PR for the fix but if someone is interested, they can investigate and submit fix.

**Comment by gaikwadrahul8** [comment]
Hi, @paradite 

I Apologize for any inconvenience this may have caused and I have re-opened this issue now and will try to reproduce the same issue from my end. Thank you!

**Comment by gaikwadrahul8** [comment]
Hi, @paradite , @POPOBE97, @nybblr 

I was trying to replicate the same issue from my end by using above code snippet provided by @POPOBE97 and I'm also getting the same error which mentioned in the above error log so we'll have to dig more into this issue and will update you soon on this issue.

Thank you for bringing this issue to our attention, I really appreciate your efforts and valuable time.

Here is error output  log for your reference :

```
gaikwadrahul-macbookpro:test-7494 gaikwadrahul$ node index.js

============================
Hi, looks like you are running TensorFlow.js in Node.js. To speed things up dramatically, install our node backend, visit https://github.com/tensorflow/tfjs-node for more details. 
============================
/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:629
            throw Error("Size(".concat(size, ") must match the product of shape ").concat(shape));
                  ^

Error: Size(4) must match the product of shape 4,4
    at Object.inferFromImplicitShape (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:629:19)
    at Object.reshape [as kernelFunc] (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-backend-cpu/dist/tf-backend-cpu.node.js:3820:32)
    at kernelFunc (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4704:32)
    at /Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4764:27
    at Engine.scopedRun (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4569:23)
    at Engine.runKernelFunc (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4760:14)
    at Engine.runKernel (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4633:21)
    at reshape_ (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs/dist/tf.node.js:4766:19)
    at reshape__op (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs/dist/tf.node.js:4024:28)
    at Object.derX [as x] (/Users/gaikwadrahul/Desktop/TFJS/test-7494/node_modules/@tensorflow/tfjs/dist/tf.node.js:8386:28)

Node.js v18.17.0
gaikwadrahul-macbookpro:test-7494 gaikwadrahul$ 
```

**Comment by paradite** [comment]
Hi @gaikwadrahul8 
I have some free time to investigate on this issue and try to submit a PR, would that be okay?

**Comment by gaikwadrahul8** [comment]
Hi, @paradite 

You're most welcome and I appreciate your willingness to help and your contribution to fixing this issue. If you have a workaround to fix this issue, please feel free to submit a PR. Our team will review your PR and take the necessary steps to merge it. Thank you!

**Comment by paradite** [comment]
Hi @gaikwadrahul8, I have submitted a PR #7942 a few days ago but not yet reviewed, can you help to get relevant people for review? 

Thanks!

**Comment by gaikwadrahul8** [comment]
Hi, @paradite 

Sure, I'll assign reviewer for your this PR https://github.com/tensorflow/tfjs/pull/7942 and they'll review your PR and will take appropriate action from their end.

Thank you for your time and submitting PR to take care of this issue, I really appreciate your valuable time and efforts. Thank you!
